Clang tidy update

mcneish1 opened the pull request:
Before:
 * 1316 warnings:
   * hicpp-vararg: 142
   * readability-redundant-string-init: 2
   * misc-string-compare: 8
   * cppcoreguidelines-pro-bounds-array-to-pointer-decay: 84
   * performance-unnecessary-value-param: 24
   * cert-err34-c: 2
   * readability-container-size-empty: 6
   * hicpp-use-auto: 53
   * clang-diagnostic-deprecated-declarations: 3
   * misc-misplaced-const: 11
   * hicpp-move-const-arg: 2
   * modernize-pass-by-value: 17
   * readability-else-after-return: 58
   * readability-simplify-boolean-expr: 8
   * hicpp-signed-bitwise: 25
   * android-cloexec-fopen: 5
   * google-runtime-int: 15
   * cppcoreguidelines-pro-type-cstyle-cast: 132
   * google-readability-casting: 33
   * performance-unnecessary-copy-initialization: 2
   * readability-implicit-bool-conversion: 117
   * readability-identifier-naming: 1
   * cppcoreguidelines-owning-memory: 126
   * clang-diagnostic-unused-parameter: 51
   * clang-diagnostic-switch: 1
   * clang-analyzer-core.uninitialized.UndefReturn: 1
   * readability-inconsistent-declaration-parameter-name: 108
   * clang-analyzer-deadcode.DeadStores: 13
   * cert-err09-cpp: 3
   * hicpp-noexcept-move: 4
   * clang-diagnostic-unused-const-variable: 1
   * hicpp-use-equals-default: 20
   * readability-redundant-member-init: 1
   * llvm-namespace-comment: 4
   * clang-analyzer-optin.cplusplus.VirtualCall: 14
   * clang-analyzer-core.CallAndMessage: 11
   * clang-diagnostic-unused-lambda-capture: 1
   * cert-dcl50-cpp: 1
   * readability-named-parameter: 75
   * modernize-use-auto: 9
   * hicpp-use-emplace: 1
   * hicpp-no-array-decay: 6
   * clang-analyzer-valist.Uninitialized: 1
   * performance-inefficient-vector-operation: 1
   * clang-diagnostic-unused-variable: 9
   * readability-non-const-parameter: 1
   * cppcoreguidelines-pro-type-member-init: 29
   * modernize-use-equals-default: 1
   * readability-delete-null-pointer: 1
   * performance-inefficient-string-concatenation: 1
   * readability-misleading-indentation: 36
   * clang-diagnostic-unreachable-code: 1
   * modernize-loop-convert: 5
   * cppcoreguidelines-pro-type-const-cast: 28
   * hicpp-member-init: 1

After (as of the time of the pr):
 * ~(clang-tidy running currently. First run segfaulted(!), but I'm not sure which file that happened on)~
 * fixed: 472
   new: 242
   (line number/offset changes mean these are both wrong, net decrease of 230)
 * Warnings: 1086
   * clang-analyzer-core.uninitialized.UndefReturn: 1
   * hicpp-use-auto: 1
   * performance-unnecessary-copy-initialization: 1
   * clang-diagnostic-unused-const-variable: 1
   * clang-diagnostic-unused-lambda-capture: 1
   * cert-dcl50-cpp: 1
   * clang-analyzer-valist.Uninitialized: 1
   * readability-non-const-parameter: 1
   * clang-diagnostic-shadow: 1
   * performance-inefficient-string-concatenation: 1
   * clang-diagnostic-unreachable-code: 1
   * hicpp-member-init: 1
   * cert-err34-c: 2
   * hicpp-move-const-arg: 2
   * google-readability-casting: 2
   * performance-unnecessary-value-param: 3
   * clang-diagnostic-deprecated-declarations: 3
   * modernize-loop-convert: 3
   * readability-misleading-indentation: 6
   * clang-diagnostic-unused-variable: 7
   * hicpp-no-array-decay: 7
   * readability-else-after-return: 8
   * clang-analyzer-optin.cplusplus.VirtualCall: 10
   * misc-misplaced-const: 11
   * clang-analyzer-core.CallAndMessage: 11
   * clang-analyzer-deadcode.DeadStores: 12
   * google-runtime-int: 14
   * modernize-pass-by-value: 14
   * hicpp-use-equals-default: 21
   * hicpp-signed-bitwise: 25
   * cppcoreguidelines-pro-type-const-cast: 28
   * cppcoreguidelines-pro-type-member-init: 31
   * clang-diagnostic-unused-parameter: 55
   * readability-named-parameter: 75
   * cppcoreguidelines-pro-bounds-array-to-pointer-decay: 85
   * readability-implicit-bool-conversion: 117
   * cppcoreguidelines-pro-type-cstyle-cast: 123
   * cppcoreguidelines-owning-memory: 126
   * readability-inconsistent-declaration-parameter-name: 131
   * hicpp-vararg: 142

Diff
 * -271, +33
   * hicpp-use-auto: -52
   * readability-else-after-return: -50
   * google-readability-casting: -31
   * readability-misleading-indentation: -30
   * performance-unnecessary-value-param: -21
   * clang-analyzer-deadcode.DeadStores: -13
   * modernize-use-auto: -9
   * cppcoreguidelines-pro-type-cstyle-cast: -9
   * misc-string-compare: -8
   * readability-simplify-boolean-expr: -8
   * readability-container-size-empty: -6
   * clang-analyzer-optin.cplusplus.VirtualCall: -4
   * llvm-namespace-comment: -4
   * hicpp-noexcept-move: -4
   * modernize-pass-by-value: -3
   * cert-err09-cpp: -3
   * readability-redundant-string-init: -2
   * clang-diagnostic-unused-variable: -2
   * modernize-loop-convert: -2
   * clang-diagnostic-switch: -1
   * google-runtime-int: -1
   * clang-analyzer-deadcode.DeadStores: -1
   * performance-unnecessary-copy-initialization: -1
   * readability-identifier-naming: -1
   * readability-redundant-member-init: -1
   * hicpp-use-emplace: -1
   * performance-inefficient-vector-operation: -1
   * modernize-use-equals-default: -1
   * readability-delete-null-pointer: -1

   * clang-diagnostic-shadow: +1
   * cppcoreguidelines-pro-bounds-array-to-pointer-decay: +1
   * hicpp-no-array-decay: +1
   * hicpp-use-equals-default: +1
   * cppcoreguidelines-pro-type-member-init: +2
   * clang-diagnostic-unused-parameter: +4
   * readability-inconsistent-declaration-parameter-name: +23

Notes:
 * Did not address yet:
   * readability-named-parameter
   * cppcoreguidelines-pro-bounds-array-to-pointer-decay
   * readability-inconsistent-declaration-parameter-name
   * readability-implicit-bool-conversion
   * cppcoreguidelines-owning-memory
   * cppcoreguidelines-pro-type-cstyle-cast
   * hicpp-vararg
   * clang-diagnostic-unused-parameter
 * Disabled android-cloexec-fopen, GridDyn never calls exec and the flag they want to add probably doesn't exist on windows
 * I have a list of TODOs on these that I'll get to next week
 * I'm going to replace all the tabs with spaces shortly: the large majority of my git conflicts were with replacing them on two different branches

mcneish1 committed:
Change spelling: data_ -> data

mcneish1 committed:
Add empty default to mode switch in fmi2Object.cpp

mcneish1 committed:
Remove unnecessary parent constructor call from dimeCommunicator

mcneish1 committed:
Use emplace_back instead of make_pair+push_back

mcneish1 committed:
Reserve appropriate vector size

mcneish1 committed:
Use '= default' for trivial destructor

mcneish1 committed:
Unconditionally delete solverKey -- delete nullptr is okay

mcneish1 committed:
Initialize std::string without '= ""'

mcneish1 committed:
Remove unused copy

mcneish1 committed:
Catch by const reference in a couple more places

mcneish1 committed:
Specify that things with &&s are noexcept

mcneish1 committed:
Fix namespace-end comments

mcneish1 committed:
Ignore cloexec warning: nothing in griddyn calls exec

mcneish1 committed:
Use range-for where applicable

mcneish1 committed:
Use .empty() instead of .size() > 0

mcneish1 committed:
Use == instead of .compare()

mcneish1 committed:
Fix a couple more .empty()s

mcneish1 committed:
Fix boolean readibility warnings

mcneish1 committed:
Use auto in more places to avoid type name duplication

mcneish1 committed:
Remove unused variables

mcneish1 committed:
Fix const gridDynObject (aka void *const) to be constGridDynObject (aka const void*)

mcneish1 committed:
Actually check for 'voltagetolerance'; don't use as many old-c-style declarations

mcneish1 committed:
Use fully-qualified member function name within constructors

mcneish1 committed:
Use std::move for by-value shared_ptrs

mcneish1 committed:
Convert most string parameters to const&, convert read-only shared_ptrs to const&, convert a read-only vector to const&, rewrite testCopyFile

mcneish1 committed:
Use c++-style casts

mcneish1 committed:
Change tabs to spaces. Change 'angleetolerance' to 'angletolerance'

mcneish1 committed:
Use auto instead of repeating templated type

mcneish1 committed:
Simplify logic following continue, else, and throw (also delete a bunch of tabs)

mcneish1 committed:
Use static_cast instead of reinterpret_cast for void*

mcneish1 committed:
Merge branches 'android-cloexec-fopen', 'cert-err09-cpp', 'clang-analyzer-deadcode--DeadStores', 'clang-analyzer-optin--cplusplus--VirtualCall', 'clang-diagnostic-switch', 'clang-diagnostic-unused-variable', 'hicpp-noexcept-move', 'hicpp-use-emplace', 'llvm-namespace-comment', 'misc-string-compare', 'modernize-loop-convert', 'modernize-pass-by-value', 'modernize-use-auto', 'modernize-use-equals-default', 'performance-inefficient-vector-operation', 'readability-delete-null-pointer', 'readability-identifier-naming' and 'readibility-redundant-member-init' into clang_tidy_update

mcneish1 committed:
Fix failure to save file after changes

mcneish1 committed:
Merge branch 'misc-misplaced-const' into clang_tidy_update

mcneish1 committed:
Merge branch 'performance-unnecessary-copy-initialization' into clang_tidy_update

mcneish1 committed:
Merge branch 'performance-unnecessary-value-param' into clang_tidy_update

mcneish1 committed:
Merge branch 'readability-container-size-empty' into clang_tidy_update

mcneish1 committed:
Merge branch 'readability-else-after-return' into clang_tidy_update

mcneish1 committed:
Merge branch 'readability-misleading-indentation' into clang_tidy_update

mcneish1 committed:
Merge branch 'readability-redundant-string-init' into clang_tidy_update

mcneish1 committed:
Merge branch 'readability-simplify-boolean-expr' into clang_tidy_update

mcneish1 committed:
Merge branch 'hicpp-use-auto' into clang_tidy_update

mcneish1 committed:
Fix compilation

 * missed a couple places where noexcept should have been
 * missed a couple const&s in headers
 * reverting the constGridDynObject thing for now
   typedefs mean it's being read as void* const instead of const void*, which is probably wrong

mcneish1 commented on .clang-tidy:
```diff
@@ -1,5 +1,5 @@
 ---
-Checks:         '*,-fuchsia-overloaded-operator,-fuchsia-default-arguments,-llvm-include-order,-llvm-header-guard,-readability-redundant-declaration,-misc-unused-parameters,-modernize-use-override,-cppcoreguidelines-pro-type-vararg,-google-default-arguments,-google-runtime-references,-google-readability-todo,-google-readability-namespace-comments,-clang-analyzer-alpha.core.CastToStruct,-clang-dia ...
+Checks:         '*,-android-cloexec-fopen,-fuchsia-overloaded-operator,-fuchsia-default-arguments,-llvm-include-order,-llvm-header-guard,-readability-redundant-declaration,-misc-unused-parameters,-modernize-use-override,-cppcoreguidelines-pro-type-vararg,-google-default-arguments,-google-runtime-references,-google-readability-todo,-google-readability-namespace-comments,-clang-analyzer-alpha.core. ...
```
https://clang.llvm.org/extra/clang-tidy/checks/android-cloexec-fopen.html

Not very applicable (we never exec) and probably doesn't work on windows

mcneish1 commented on src/fileInput/readerInfo.cpp:
```diff
@@ -193,32 +193,29 @@ std::string readerInfo::checkDefines (const std::string &input)
                     break;  // break out of inner loop
                 }
-                else
+                // try a recursive interpretation of the string block to convert a numeric result back into a
```
This should just be decreasing a nesting level with the else removal

mcneish1 commented on src/fmi/fmi_models/fmiMESubModel.cpp:
```diff
@@ -370,75 +357,45 @@ void fmiMESubModel::set (const std::string &param, double val, gridUnits::units_
 {
 	stateSizes SS;
-	if (hasDifferential(sMode))
+	if (hasDifferential(sMode) or (!isDynamic(sMode) and opFlags[pflow_init_required]))
```
this is wrong, fixing now

mcneish1 replied:
(didn't catch diffSize vs algSize)

mcneish1 commented on src/fmi/fmi_models/outputEstimator.cpp:
```diff
@@ -21,10 +21,8 @@ namespace griddyn
 {
 
-outputEstimator::outputEstimator(std::vector<int> sDep, std::vector<int> iDep)
+outputEstimator::outputEstimator(std::vector<int> sDep, std::vector<int> iDep) : stateDep(std::move(sDep)), inputDep(iDep)
```
Should probably move iDep too

mcneish1 commented on src/fmi_export/fmiCollector.cpp:
```diff
@@ -41,13 +41,6 @@ std::unique_ptr<collector> fmiCollector::clone() const
 {
 	collector::cloneTo(gr);
-
-	auto nrec = dynamic_cast<fmiCollector *>(gr);
```
Was something supposed to happen to nrec?

phlptp replied:
I don't think the function was finished.  The fmi stuff is still in progress, so some parts are not finished.  so yes but it hasn't been written yes, and I am not sure of a use case for the function yet just do the nature of the situations in which this is used.

phlptp replied:
I am guessing I just put the boilerplate code in place

mcneish1 replied:
👍

mcneish1 commented on src/fmi_export/fmuBuilder.cpp:
```diff
@@ -108,34 +108,20 @@ int fmuBuilder::Initialize (int argc, char *argv[])
-        {
-            return false;
-        }
+        copy_file (source, dest, option);
```
copy_file does existence/overwrite handling already: just give it the right `copy_option`

mcneish1 commented on src/gridDynMain/CMakeLists.txt:
```diff
@@ -8,30 +8,30 @@
 # LLNS Copyright End
 
-set(griddynMain_sources
+set(gridDynMain_sources
```
Capitalization? It's spelled this way in the manual

phlptp replied:
ok

mcneish1 commented on src/griddyn/comms/commManager.cpp:
```diff
@@ -35,7 +35,8 @@ commManager::commManager (const commManager &cm)
 }
 
-commManager::commManager (commManager &&) = default;
+commManager::commManager (commManager &&) noexcept = default;
```
Move constructors generally need to be noexcept to be selected. I didn't actually check to see if any of the things I marked could throw, though

phlptp replied:
need to check whether the header marked it as noexcept.  I think the =default will mark it as noexcept

mcneish1 replied:
https://stackoverflow.com/a/18654089 looks like you're right. I wonder why it warned me then

mcneish1 commented on src/griddyn/loads/aggregateLoad.cpp:
```diff
@@ -114,14 +114,16 @@ void aggregateLoad::pFlowObjectInitializeA (coreTime time0, std::uint32_t flags)
 
-        setP (((aggregateLoad *)sLoad)->getP ());  // so this composite load function has direct access to the
+        aggregateLoad* sLoadRepr = dynamic_cast<aggregateLoad*>(sLoad);
+        assert(sLoadRepr != nullptr);
```
Need to look at zipLoad vs aggregateLode, to see if this can actually succeed

phlptp replied:
aggregate load builds on zipload, so all aggregate loads are ziploads.

mcneish1 commented on src/griddyn/loads/motorLoad5.cpp:
```diff
@@ -83,27 +83,24 @@ void motorLoad5::converge ()
     double Qtest = qPower (V, m_state[2]);
```
More of these are unused loop-to-loop

mcneish1 commented on src/griddyn/loads/svd.cpp:
```diff
@@ -91,71 +91,67 @@ int svd::checkSetting (double level)
-		double totalQ = Qlow;
-		if (!opFlags[reverse_control_flag])
+
+    int setting = 0;
```
TODO the indentation is wrong here

mcneish1 replied:
Just kidding it's tabs vs spaces. Not fixing yet

phlptp commented on src/fmi/fmi_import/fmiInfo.cpp:
```diff
@@ -121,22 +121,18 @@ int fmiInfo::getCounts(const std::string &countType) const
 }
 
-const std::string emptyString="";
+const std::string emptyString;
```
This should probably be static const

mcneish1 commented on src/griddyn/relays/Relay.cpp:
```diff
@@ -840,7 +837,7 @@ std::unique_ptr<eventAdapter> Relay::make_alarm (const std::string &val)
 }
 
-void Relay::receiveMessage (std::uint64_t /*sourceID*/, std::shared_ptr<commMessage> /*message*/) {}
+void Relay::receiveMessage (std::uint64_t sourceID, std::shared_ptr<commMessage> const& message) {}
```
Unused parameters are going to be in a future commit

mcneish1 commented on src/griddyn/solvers/arkodeInterface.cpp:
```diff
@@ -324,7 +324,7 @@ void arkodeInterface::initialize (coreTime time0)
     // dynInitializeB CVode - Sundials
 
-    int retval = ARKodeSetUserData (solverMem, reinterpret_cast<void *> (this));
+    int retval = ARKodeSetUserData (solverMem, static_cast<void *> (this));
```
I _think_ these only need to be static and not reinterpret

phlptp replied:
in reading a little more into this static_cast can cast to void *.  in a few cases things need to be cast to char * or const char *.  this would require the reinterpret_cast<char *>.

mcneish1 commented on src/griddyn/solvers/sundialsInterface.cpp:
```diff
@@ -1,532 +1,534 @@
-/*
```
A lot of things happened to this file. I'm tempted to restart from the previous version in cmake_update, just so the git history actually makes sense

mcneish1 replied:
I'm pretty sure the only 2 real changes are tabs->spaces and static_cast<void*> instead of c-style casts

phlptp commented on src/griddyn/governors/GovernorSteamNR.cpp:
```diff
@@ -75,7 +75,7 @@ void GovernorSteamNR::dynObjectInitializeB (const IOdata & /*inputs*/,
     m_state[offset + offset] = desiredOutput[0];
 
-    Pset = ((Generator *)getParent ())->getPset ();
+    Pset = dynamic_cast<Generator*> (getParent ())->getPset ();
```
this should be static_cast,  it is guaranteed that the parent is a generator by construction if it gets here

mcneish1 replied:
```
coreObject *GovernorSteamNR::clone (coreObject *obj) const
{
    GovernorSteamNR *gov;
    if (obj == nullptr)
    {
        gov = new GovernorSteamNR ();
    }
    else
    {
        gov = dynamic_cast<GovernorSteamNR *> (obj);
        if (gov == nullptr)
        {
            coreObject::clone (obj);
            return obj;
        }
        ...
}
```
should be
```
coreObject *GovernorSteamNR::clone (coreObject *obj) const
{
    GovernorSteamNR *gov;
    if (obj == nullptr)
    {
        gov = new GovernorSteamNR ();
    }
    else
    {
        gov = static_cast<GovernorSteamNR *> (obj);
        ...
}
```
then? static_cast<T>(obj) won't be null unless obj is

phlptp replied:
This class is not operational, it will be rewritten at a later time.  The model is wrong, and the code is an old copied placeholder.  I am waiting to get a more generic control system model running, before replacing it though.

mcneish1 commented on src/optimization/models/gridBusOpt.cpp:
```diff
@@ -535,216 +535,192 @@ void gridBusOpt::remove (gridLinkOpt *lnk)
 
     }
-  else if ((param == "angleetolerance") || (param == "atol"))
+    else if ((param == "angletolerance") || (param == "atol"))
```
Fixed all of these I found. I haven't checked other strings for typos very thoroughly yet

mcneish1 commented on src/optimization/models/gridRelayOpt.cpp:
```diff
@@ -34,115 +34,111 @@ gridRelayOpt::gridRelayOpt (const std::string &objName) : gridOptObject (objName
 {
-  count_t objs = 0;
-  switch (oMode.linMode)
+    /*
```
This looked unfinished too

phlptp replied:
Most of the optimization library is unfinished

phlptp commented on src/griddyn/loads/aggregateLoad.cpp:
```diff
@@ -114,14 +114,16 @@ void aggregateLoad::pFlowObjectInitializeA (coreTime time0, std::uint32_t flags)
         double rem = 1.0;
 
-        setP (((aggregateLoad *)sLoad)->getP ());  // so this composite load function has direct access to the
+        aggregateLoad* sLoadRepr = dynamic_cast<aggregateLoad*>(sLoad);
```
this is a bit of weird code.  You need to use reinterpret_cast here,  dynamic_cast won't work properly.  We know the object is a zipload, but for the compiler to allow these calls which are purely part of the zipload interface it has to think it is a aggregate load.  
We probably should find a different way of doing this but for now it needs to be reinterpret_cast not dynamic_cast

phlptp commented on src/griddyn/loads/aggregateLoad.cpp:
```diff
@@ -189,11 +191,9 @@ void aggregateLoad::dynObjectInitializeA (coreTime time0, std::uint32_t flags)
     for (auto &ld : subLoads)
     {
-        ld->dynInitializeB (inputs, out, fieldSet);
+        ld->dynInitializeB (inputs, desiredOutput, fieldSet);
```
Add a note here that more needs to be done in case desiredOutput actually contains something of use.  I believe all examples use it as empty so this won't matter, but eventually something needs to be one.  So we should add a TODO here

phlptp commented on src/griddyn/relays/controlRelay.h:
```diff
@@ -93,15 +93,15 @@ class controlRelay : public Relay
   virtual void actionTaken (index_t ActionNum, index_t conditionNum, change_code actionReturn, coreTime actionTime) override;
 
-  virtual void receiveMessage (std::uint64_t sourceID, std::shared_ptr<commMessage> message) override;
+  virtual void receiveMessage (std::uint64_t sourceID, std::shared_ptr<commMessage> const& message) override;
```
This is a virtual function we would need to make sure all uses are changed likewise and are amenable to the change,  which given this can be called in a multithread situation I am not sure I am comfortable passing around a reference to a shared pointer.

mcneish1 replied:
I changed all uses. Whatever calls receiveMessage will have ownership of message, so there should always be at least one reference, right? I'm not sure how it could be destructed even with threads.

phlptp commented on src/griddyn/solvers/solverInterface.cpp:
```diff
@@ -571,10 +571,10 @@ void SolverInterface::check_flag (void *flagvalue, const std::string &funcname,
     {
         // Check if flag < 0
-        errflag = reinterpret_cast<int *> (flagvalue);
+        errflag = static_cast<int *> (flagvalue);
```
I think this one needs to be reinterpret_cast

phlptp reviewed (changes requested):
made a few comments.  A few of the casts look wrong

mcneish1 commented:
- [ ] src/fmi_export/fmiCollector.cpp, src/griddyn/governors/GovernorSteamNR.cpp, and src/optimization/models/gridRelayOpt.cpp -- revert, since unused/unfinished
 - [ ] zipLoad reinterpret_cast
 - [ ] look through the casts again
 - [ ] noexcept = default; -- revert, works on clang but not travis' gcc
 - [ ] desiredOutput -- add comment on use
 - [ ] clean up and add clang-tidy output parser

mcneish1 committed:
Use static for emptyString; fix LocalStateSizes logic; std::move additional vector

mcneish1 committed:
Revert noexcept additions

mcneish1 committed:
Revert changes to unfinished files

mcneish1 committed:
Update executable name in ci script

mcneish1 committed:
Fix other gridDynMain spellings

mcneish1 committed:
Remove a couple more noexcept

mcneish1 committed:
This prints:
x: 0
x: 5

class test_base
{

protected:
    double x = 0;
    void set_x(double y) { x = y; }
    double get_x() { return x; }

};

class test_derived : public test_base
{
public:
    void print_x()
    {
        std::cout << "x: " << get_x() << std::endl;
        set_x(5);
        std::cout << "x: " << get_x() << std::endl;
    }
};

int main()
{
    test_derived a;
    a.print_x();
    return 0;
}

mcneish1 committed:
void* casts are implicit

mcneish1 committed:
reinterpret_cast the sundials return value if necessary

mcneish1 committed:
Add clang-tidy ease-of-use scripts

mcneish1 committed:
Set eventId in initializer list

mcneish1 committed:
Use an external file for specifying the current guideline

mcneish1 committed:
void* conversion is implicit

phlptp commented on src/fmi/fmi_import/fmiInfo.h:
```diff
@@ -114,10 +113,10 @@ class fmiVariableSet
 	fmiVariableSet(fmi2ValueReference newvr);
 	fmiVariableSet(const fmiVariableSet &vset);
-	fmiVariableSet(fmiVariableSet &&vset) noexcept;
+	fmiVariableSet(fmiVariableSet &&vset);
```
I think this could be noexcept in the declaration,  but then you wouldn't use the noexcept in the instantiation

mcneish1 replied:
Yeah that was a mistake

To clarify, this commit fixes the partial revert of a commit that broke all of the Travis compilers. I'm trying to finish up this pull request, I'll look more into why it was an issue in the first place a little later.

phlptp commented on src/fmi_export/fmiCollector.cpp:
```diff
@@ -52,15 +59,23 @@ change_code fmiCollector::trigger(coreTime time)
 void fmiCollector::set(const std::string &param, double val)
 {
-	if (param.front() != '#')
+	if (param.front() == '#')
```
This is should really be `if (param.empty())`  if param is empty this will cause a seg fault

phlptp replied:
there are probably other places this shows up that should be changed as well

mcneish1 replied:
There are a lot of places where params are getting parsed. I have them down as something to look over, but that's going to be a little later. (This diff is just a checkout of the file from cmake_update)

phlptp commented on src/optimization/optimizerInterface.cpp:
```diff
@@ -49,7 +49,6 @@ double optimizerInterface::get (const std::string & /*param*/) const
 
 int optimizerInterface::check_flag (void *flagvalue, const std::string &funcname, int opt, bool printError)
```
this function is a replicate of what was in solverInterfaces,  it probably isn't needed here since we are updating the optimizer interface.

phlptp commented on tidy/clang_tidy.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
you may want to move these files to the scripts folder and look at using/updating the cmake targets that do some of this already

mcneish1 replied:
👍

mcneish1 committed:
Use builtin filter; note that desiredOutput might need to be used

mcneish1 committed:
Move scripts to scripts/

mcneish1 commented:
I think the only thing left to do is decide whether you want the receiveMessage modifications

testSystemQuick in travis segfaulted on most of my other pushes, but only for gcc4.9. Not sure what's going on there

phlptp commented:
I think my inclination is to leave it as pass by value,  The function name itself implies ownership transfer, and I tend to think passing a shared_ptr by reference is a bit odd to begin with,  should probably be by pointer or just object reference at that point.    My guess is there are cases where the calling function is actually moving a shared_ptr or unique_ptr into the function, or it is actually intending to share a reference.  Thus in most cases if you have it as pass by reference it would cause an extra temporary to be created anyway.  

gcc 4.9.4 has some odd behavior with shared_ptr arguments particularly when translating from derived class pointers.  

I think it would be best to leave it as it was with passing a shared_ptr by value

mcneish1 committed:
Remove const&

phlptp commented:
Do we have any idea where the segfault occurs?  I seem to recall that happening before as well, and not being able to track it down.

mcneish1 commented:
Errors I can reproduce:
(all 4 of these are using the same suite-sparse, but self-built boost/sundials. all time out on testExtra)
 * gcc4.9.3
   * ```testSystem: GridDyn/src/griddyn/primary/acBus.cpp:1963: virtual void griddyn::acBus::residual(const IOdata&, const griddyn::stateData&, double*, const griddyn::solverMode&): Assertion `!std::isnan (S.linkQ)' failed.
unknown location(0): fatal error: in "fault_tests/geco_fault_case": signal: SIGABRT (application abort requested)```
 * gcc7.3.0
   * ```testSystem: GridDyn/src/griddyn/primary/acBus.cpp:1963: virtual void griddyn::acBus::residual(const IOdata&, const griddyn::stateData&, double*, const griddyn::solverMode&): Assertion `!std::isnan (S.linkQ)' failed.
unknown location(0): fatal error: in "fault_tests/geco_fault_case": signal: SIGABRT (application abort requested)```
 * clang6
   * no errors with what's enabled by default
   * error with most things enabled:
```
unknown location(0): fatal error: in "fmi_export_tests/test_fmi_runner": std::runtime_error: GhostSwingBusManager Instance does not exist!
GridDyn/src/fmi_export/testFMIExport.cpp(135): last checkpoint
GridDyn version:0.9.0 4-28-2018
area count =0 buses=2 links= 1 gens= 0
fmu created at GridDyn/test/test_files/fmi_export_tests/griddyn.fmu
GridDyn/src/fmi_export/testFMIExport.cpp(170): fatal error: in "fmi_export_tests/load_griddyn_fmu": critical check gdFmu.isSoLoaded () has failed
unknown location(0): fatal error: in "fmi_export_tests/test_fmi_runner2": memory access violation at address: 0x00000000: no mapping at fault address
GridDyn/src/fmi_export/testFMIExport.cpp(253): last checkpoint: "test_fmi_runner2" test entry```

mcneish1 commented:
I don't think things got worse with this pr, and I'd like to do the investigation after I fix all the formatting. Is there anything about this one that you still have problems with?

phlptp commented:
testExtra is not for regular execution, so that it timed out doesn't really surprise me.

mcneish1 commented:
running testExtra manually:

```
Pure virtual function called!Aborted (core dumped)
[from the core dump]
#0  0x00002aaaad08f207 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00002aaaad0908f8 in __GI_abort () at abort.c:90
#2  0x00002aaaac714b6c in abort_message () from /[...]/lib/libc++abi.so.1
#3  0x00002aaaac714512 in __cxa_pure_virtual () from /[...]/lib/libc++abi.so.1
#4  0x0000000000c19bce in griddyn::acBus* griddyn::cloneBaseFactory<griddyn::acBus, griddyn::gridBus>(griddyn::acBus const*, griddyn::coreObject*, griddyn::objectFactory*) ()
#5  0x0000000000bf5ad2 in griddyn::acBus::clone(griddyn::coreObject*) const ()
#6  0x0000000000bb6502 in griddyn::Area::clone(griddyn::coreObject*) const ()
#7  0x000000000079bd97 in griddyn::gridSimulation* griddyn::cloneBase<griddyn::gridSimulation, griddyn::Area>(griddyn::gridSimulation const*, griddyn::coreObject*) ()
#8  0x000000000078b6a2 in griddyn::gridSimulation::clone(griddyn::coreObject*) const ()
#9  0x0000000000749b97 in griddyn::gridDynSimulation* griddyn::cloneBase<griddyn::gridDynSimulation, griddyn::gridSimulation>(griddyn::gridDynSimulation const*, griddyn::coreObject*) ()
#10 0x000000000072c157 in griddyn::gridDynSimulation::clone(griddyn::coreObject*) const ()
#11 0x000000000076808e in griddyn::contingency::execute() ()
#12 0x00000000011575c6 in workQueue::workerLoop() ()
#13 0x0000000001160c43 in void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (workQueue::*)(), workQueue*> >(void*) ()
#14 0x00002aaaace44dd5 in start_thread (arg=0x2aaaba3ba700) at pthread_create.c:308
#15 0x00002aaaad157b3d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:113
```

phlptp commented:
The testSystemQuick on Gcc 4.9 is failing with a segfault.  Do we know where the segfault occurs?

mcneish1 commented:
That's what I addressed [here](https://github.com/LLNL/GridDyn/pull/22#issuecomment-404258093), I can reproduce some other things but not that segfault.

```
[mcneish1@borax5:build_segfault]$ ./test/testSystem --run_test=@quick
Running 62 test cases...
(0.000000)[sim]::||WARNING||solver error return code:-11
completed power flow
Unable to locate main executable:: skipping test
(3.000000)[bus4(225)]::bus low Voltage
(3.000000)[sim]::||WARNING||solver error return code:-6
(3.000000)[bus4(225)]::bus low Voltage
(3.000000)[sim]::||WARNING||solver error return code:-6

*** No errors detected
```
This is with boost 1.67.0, gcc 4.9.3, and suite-sparse 5.2.0 preinstalled, I'm trying with the exact dependencies Travis uses now.

mcneish1 commented:
Okay, I had GridDyn build SuiteSparse and used the script to install boost. I get this:

```
testSystem: GridDyn/src/griddyn/primary/acBus.cpp:1963: virtual void griddyn::acBus::residual(const IOdata&, const griddyn::stateData&, doub
le*, const griddyn::solverMode&): Assertion `!std::isnan (S.linkQ)' failed.
unknown location(0): fatal error: in "fault_tests/geco_fault_case": signal: SIGABRT (application abort requested)
GridDyn/test/systemTests/faultTests.cpp(212): last checkpoint
```

but still no segfault

phlptp commented:
@mcneish1 I think I made a conflict in one of my recent commits.  you might need to fix it.  Is there anything else you are planning to add to this commit?

mcneish1 commented:
I saw that. The only non-formatting change is 
```
+    if (scale != nullptr)
+    {
+        NVECTOR_DESTROY (use_omp, scale);
+    }
+    if (types != nullptr)
+    {
+        NVECTOR_DESTROY (use_omp, types);
+    }
```
right?

Yeah, this one's done after that. I'm tracking down some of the failures but they've been around [awhile](https://travis-ci.org/LLNL/GridDyn/jobs/390993791), my changes shouldn't have made them any worse

phlptp commented:
I think those were the substantive changes.  I think they fix some memory leaks.  If you resolve the conflict. I will take a last look through it and merge it if I don't find any new issues.

mcneish1 committed:
Free a couple more things in the sundials interface

mcneish1 committed:
Merge branch 'cmake_update' into clang_tidy_update

Use my version, already integrated the non-formatting change

mcneish1 commented:
Added the fixes then just took my version

phlptp commented on src/griddyn/loads/aggregateLoad.cpp:
```diff
@@ -114,15 +114,6 @@ void aggregateLoad::pFlowObjectInitializeA (coreTime time0, std::uint32_t flags)
-        setIq (((aggregateLoad *)sLoad)->getIq ());
-        setYp (((aggregateLoad *)sLoad)->getYp ());
-        setYq (((aggregateLoad *)sLoad)->getYq ());
-
```
Why was this removed?  I think it is still necessary?

mcneish1 replied:
Reading it again, you're right. I thought sLoad was `this`, but I'm not sure why. Going to revert this

phlptp commented on src/griddyn/solvers/solverInterface.cpp:
```diff
@@ -571,10 +571,10 @@ void SolverInterface::check_flag (void *flagvalue, const std::string &funcname,
     {
         // Check if flag < 0
-        errflag = reinterpret_cast<int *> (flagvalue);
+        int *errflag = reinterpret_cast<int *> (flagvalue);
```
should be auto

phlptp commented on src/networking/TcpHelperClasses.cpp:
```diff
@@ -29,10 +29,7 @@ void TcpRxConnection::start ()
     connection_state_t exp = connection_state_t::halted;
     if (state.compare_exchange_strong (exp, connection_state_t::receiving))
     {
-        if (receiving == false)
```
this probably should be left as it was.  The reason for this is because receiving is an atomic and reads are alot less expensive then writes so we don't want to do a write unless we really need to.

phlptp commented:
@mcneish1   a few more changes, then I think it is good to go.  I made comments on the specific changes

mcneish1 committed:
Fix remaining issues
